pkg/resource/deploy/snapshot.go

```diff
@@ -49,6 +49,99 @@ func NewSnapshot(manifest Manifest, secretsManager secrets.Manager,
+	// states, and so will have no need to refer to the clobbered entries.
+	seen := map[resource.URN]resource.URN{}
+
+	for _, state := range snap.Resources {
+		newURN := state.URN
+
+		newDeps := []resource.URN{}
+		newPropDeps := map[resource.PropertyKey][]resource.URN{}
+
+		// If a provider reference is dangling, there's not much we can do -- resource states *must* have providers, so we
+		// can't simply remove it. Better to leave it so that VerifyIntegrity can spot it and present an appropriate error
+		// message.
+		_, allDeps := state.GetAllDependencies()
+		for _, dep := range allDeps {
+			switch dep.Type {
+			case resource.ResourceParent:
+				// Since parent-child relationships affect URNs, we have more work to do for a parent dependency. If our parent
+				// is missing, we'll clear the reference and update our URN to remove the parent type. Moreover, we'll record
+				// the fact that we rewrote our URN so that any of our children can update their URNs appropriately.
+				//
+				// If our parent is present, but was rewritten, we'll need to rewrite our URN and record that it was rewritten
+				// for our children, and so on.
+				//
+				// Note: the precondition that the snapshot is topologically sorted allows us to assume that our parent's
+				// presence/rewriting has already been determined.
+				newParentURN, has := seen[dep.URN]
+				if !has {
+					newURN = resource.NewURN(state.URN.Stack(), state.URN.Project(), "", state.URN.Type(), state.URN.Name())
+					state.Parent = ""
+				} else {
+					newURN = resource.NewURN(
+						state.URN.Stack(),
+						state.URN.Project(),
+						newParentURN.QualifiedType(),
+						state.URN.Type(),
+						state.URN.Name(),
+					)
+					state.Parent = newParentURN
+				}
+			case resource.ResourceDependency:
+				// For dependencies, only preserve those that aren't dangling, taking into account any rewrites that may have
+				// occurred.
+				if newDepURN, has := seen[dep.URN]; has {
+					newDeps = append(newDeps, newDepURN)
+				}
+			case resource.ResourcePropertyDependency:
+				// For property dependencies, only preserve those that aren't dangling, taking into account any rewrites that
+				// may have occurred.
+				if newPropDepURN, has := seen[dep.URN]; has {
+					newPropDeps[dep.Key] = append(newPropDeps[dep.Key], newPropDepURN)
+				}
+			case resource.ResourceDeletedWith:
+				// Only preseve a deleted-with relationship if it isn't dangling, taking into account any rewrites that may have
+				// occurred.
+				if newDeletedWithURN, has := seen[dep.URN]; has {
+					state.DeletedWith = newDeletedWithURN
+				} else {
+					state.DeletedWith = ""
+				}
+			}
+		}
+
+		// Since we can only have shrunk the sets of dependencies and property dependencies, we'll only update them if they
+		// were non-empty to begin with. This is to avoid e.g. replacing a nil input with an non-nil but empty output, which
```

Comment:
Do we want to replace new empty sets _with_ nil?

Reply:
Given that this is only really going to be used in one place initially (`state repair`), we could (assuming you are suggesting that we _should_ pare empty sets to `nil`). That said, I've taken this approach since it matches `rebuildBaseState`'s behaviour in response to a deleting `refresh`, which I believe was done originally because it does have subtle affects on how diffs render. Happy either way if you have a preference?

Reply:
I don't have a preference, I think anywhere we treat nil different to [] is a bug asking to happen.